core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -979,6 +1004,21 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         }
     }
 
+    private void verifyPrimary() {
+        if (shardRouting.primary() == false) {
+            // must use exception that is not ignore be replication logic. See TransportActions.isShardNotAvailableException
+            throw new UnsupportedOperationException("shard is not a primary " + shardRouting);
```

Comment:
I wonder how this change in exception type (from exception which leads to retry to exception that does not enable retry) is reflected in the tests. Do we have tests that check the behavior we now want to have?